python/ray/data/_internal/logical/rules/inherit_batch_format.py

```diff
@@ -0,0 +1,37 @@
+from collections import deque
+
+from ray.data._internal.logical.interfaces import LogicalOperator, LogicalPlan, Rule
+from ray.data._internal.logical.operators.all_to_all_operator import Sort
+from ray.data._internal.logical.operators.map_operator import MapBatches
+
+
+class InheritBatchFormatRule(Rule):
+    # TODO(xingyu-long): add doc here
+
+    def apply(self, plan: LogicalPlan) -> LogicalPlan:
+        optimized_dag: LogicalOperator = self._apply(plan.dag)
+        new_plan = LogicalPlan(dag=optimized_dag, context=plan.context)
+        return new_plan
+
+    def _apply(self, op: LogicalOperator):
+        # Post-order traversal.
+        nodes = deque()
+        for node in op.post_order_iter():
+            nodes.appendleft(node)
+
+        while len(nodes) > 0:
+            current_op = nodes.pop()
+            upstream_ops = current_op.input_dependencies
+
+            # TODO(xingyu-long): extend this to other LogicalOperator as well
+            # e.g., shuffle, aggregate
+            if isinstance(current_op, Sort):
+                batch_format = None
+                for i in range(len(upstream_ops)):
+                    if isinstance(upstream_ops[i], MapBatches):
+                        if upstream_ops[i]._batch_format is not None:
```

Comment:
i believe `MapBatches._batch_format` will always be a string. so we can directly inherit the value instead of checking for `None`.

however, one case i think we need to consider here is if the `MapBatches` op is not directly upstream of `Sort`. for example, consider the following plan: `MapBatches->Filter->Sort`. in this case, we would still want to inherit the `batch_format` from `MapBatches`, because `Filter` will be a row-wise operation which doesn't have any batch format associated.

so, if none of `upstream_ops` contains `MapBatches`, we should traverse up the DAG until we either find one, or just use the default batch format if we reach the source op (in this case, i think we don't need to set batch_format at all, since it's being set in `Sort.__init__()`).

Reply:
Yeah, I made it more general by traversing the upstream path. Let me know if it works.